src/fourier.jl

```diff
@@ -0,0 +1,755 @@
+    minbin = 2
+    if include_zero
+        minbin = 1
+    end
+
+    if n_bin%2 == 0
+        return (minbin : floor(Int, n_bin / 2))
+    end
+    return (minbin : floor(Int, (n_bin+1)/ 2))
+end
+
+function poisson_level(;norm = "frac", meanrate = nothing, n_ph = nothing, backrate = 0)
+    if norm == "abs"
+        return 2.0 * meanrate
+    elseif norm == "frac"
+        return 2.0 / (meanrate - backrate)^2 * meanrate
+    elseif norm == "leahy"
+        return 2.0
+    elseif norm == "none"
+        return float(n_ph)
+    end
+end
+
+function normalize_frac(unnorm_power, dt, n_bin, mean_flux; background_flux=0.0)
+    if background_flux > 0
+        power = unnorm_power * 2. * dt / ((mean_flux - background_flux) ^ 2 *
+                                          n_bin)
+    else
+        # Note: this corresponds to eq. 3 in Uttley+14
+        power = unnorm_power * 2. * dt / (mean_flux ^ 2 * n_bin)
+    end
+    return power
+end
+
+function normalize_abs(unnorm_power, dt, n_bin)
+    return unnorm_power * 2.0 / n_bin / dt
+end
+
+function normalize_leahy_from_variance(unnorm_power, variance, n_bin)
+    return unnorm_power * 2.0 / (variance * n_bin)
+end
+
+function normalize_leahy_poisson(unnorm_power, n_ph)
+    return unnorm_power * 2.0 / n_ph
+end
+
+function normalize_periodograms(unnorm_power, dt, n_bin; mean_flux=nothing, n_ph=nothing,
+    variance=nothing, background_flux=0.0, norm="frac",
+    power_type="all")
+    
+    if norm == "leahy" && !isnothing(variance)
+        pds = normalize_leahy_from_variance(unnorm_power, variance, n_bin)
+    elseif norm == "leahy"
+        pds = normalize_leahy_poisson(unnorm_power, n_ph)
+    elseif norm == "frac"
+        pds = normalize_frac(
+            unnorm_power, dt, n_bin, mean_flux,
+            background_flux=background_flux)
+    elseif norm == "abs"
+        pds = normalize_abs(unnorm_power, dt, n_bin)
+    elseif norm == "none"
+        pds = unnorm_power
+    end
+
+    if power_type == "all"
+        return pds
+    elseif power_type == "real"
+        return real(pds)
+    elseif power_type in ["abs", "absolute"]
+        return abs.(pds)
+    end
+
+end
+
+function bias_term(power1, power2, power1_noise, power2_noise, n_ave;
+                   intrinsic_coherence=1.0)
+    
+    if n_ave > 500
+        return 0.0 .* power1
+    end
+    bsq = power1 .* power2 - intrinsic_coherence * (power1 .- power1_noise) .* (power2 .- power2_noise)
+    return bsq / n_ave
+end
+
+function raw_coherence(cross_power, power1, power2, power1_noise, power2_noise,
+                       n_ave; intrinsic_coherence=1.0)
+    bsq = bias_term(power1, power2, power1_noise, power2_noise, n_ave;
+                    intrinsic_coherence=intrinsic_coherence)
+    num = real(cross_power .* conj(cross_power)) - bsq
+    if num isa Array{T} where T<:Any
+        num[num .< 0] = real(cross_power .* conj(cross_power))[num .< 0]
+    elseif num < 0
+        num = real(cross_power * conj(cross_power))
```

Comment:
Are you sure this line is doing what you think?  Please elaborate in words what's your expectation.

Reply:
I can't use dot broadcast. The num .<0 returns a boolean array (suppose B). B[i] is false if num[i]<0 and num[B] gives a new array only consisting of those elements of num where B is true. (I think filter could be used to create the array, but dot broadcasting not in my opinion).

Reply:
My point is that you can define a function (please use a more sensible name than "foo")
```julia
foo(num::Real) = num < 0 ? real(cross_power * conj(cross_power))
```
and then here you do
```julia
num .= foo.(num)
```
which would work whatever `num` is, either a scalar number or an aray.  Much simpler, and it doesn't allocate lots of temporary new arrays.

Reply:
Problem is cross power is also an array, so this way won't work. If we have to use function we need to pass index too, in that case also, there is a separation between array and number. num is simply iterable of floats (I think it works with tuples too in python)

Reply:
Why it wouldn't work?

Reply:
```
julia> cross_power = [1+im,1+2im,2]
3-element Vector{Complex{Int64}}:
 1 + 1im
 1 + 2im
 2 + 0im

julia> num=[-1,-2,3]
3-element Vector{Int64}:
 -1
 -2
  3

julia> foo(num::Real) = num < 0 ? real(cross_power .* conj(cross_power)) : num   
foo (generic function with 1 method)

julia> num .= foo.(num)
ERROR: MethodError: Cannot `convert` an object of type Vector{Int64} to an object of type Int64
```

Reply:
```julia
julia> foo(num::Real, cross_power::Complex) = num < 0 ? real(cross_power * conj(cross_power)) : num
foo (generic function with 1 method)

julia> cross_power = [1 + im, 1 + 2im, 2]
3-element Vector{Complex{Int64}}:
 1 + 1im
 1 + 2im
 2 + 0im

julia> num = [-1.0, -2.0, 3.0]
3-element Vector{Float64}:
 -1.0
 -2.0
  3.0

julia> num .= foo.(num, cross_power)
3-element Vector{Float64}:
 2.0
 5.0
 3.0
```

Reply:
Oh yes we can do that, I looked now upon different ways to broadcast arguments in functions, nice!

Reply:
Now this is working correctly!